src/probability/bells_inequality.lean

```diff
@@ -0,0 +1,180 @@
+Here, the inequality is formulated in terms of probabilities.
+
+Bell's inequality is the cornerstone of Bell's theorem, which states that
+quantum mechanics is a non-local theory. The proof of Bell's theorem,
+established by J.S. Bell in 1964, first uses the Einstein-Podolsky-Rosen
+argument to establish that either quantum mechanics is non-local, or all
+quantum observables simultaneously have values (in other words, quantum
+mechanics is a non-contextual hidden variable theory). Bell's inequality is a
+necessary condition for all the observables in an abstract probabilistic theory
+to simultaneously have values. By showing that, in quantum mechanics, Bell's
+inequality is violated, Bell proved that quantum mechanics is non-local.
+
+The violation of Bell's inequality has also been verified experimentally in
+many different setups.
+
+
+## Bell's 1964 inequality
+
+We first prove Bell's original statement of the inequality, which was published
+in 1964, and will thus be called `bells_inequality_1964` in this file.
+
+Consider two observers, A and B, performing measurements that take values in
+{-1,+1}.
+
+Let `ℙ` be a probability measure, and let `Za i` and `Zb i` for i∈{1,2,3} be
+random variables taking values in {-1,+1}. For convenience, we identify {-1,+1}
+with ℤˣ. `Za i` represent the outcomes of measurements done by A, and `Zb i`
+those done by B. We assume perfect anticorrelation between the outcomes of A
+and B: 𝔼[(Za i) (Zb i)] = -1. Bell's inequality states that
+  `𝔼[(Za 1) (Zb 2)] - 𝔼[(Za 1) (Zb 2)] ≤ 1 + 𝔼[(Za 2) (Zb 3)]`.
+
+
+## Future work
+
+J.S. Bell generalized the inequality in 1975 to include more probabilistic
+theories. In particular, the 1975 statement does not require observables to
+simultaneously have values. Instead, it is solely based on the requirement of
+locality. The 1975 inequality thus gives a more direct proof of Bell's theorem.
+
+
+## References
+
+* [J.S. Bell, *On the Einstein Podolsky Rosen Paradox*, 1964][MR3790629]
+
+* [J.S. Bell, *The theory of local beables*, 1975,
+  reproduced in chapter 7 of *Speakable and unspeakable in quantum mechanics*][MR915338]
+
+
+
+-/
+
+noncomputable theory
+
+open measure_theory
+
+section preliminaries
+
+lemma pm_one_space_vals (r : ℤˣ) :
+  (r : ℝ) = 1 ∨ (r : ℝ) = -1 := by cases int.units_eq_one_or r with hh hh; rw hh; simp
+
+lemma pm_one_space_abs_le (r : ℤˣ) :
+  |(r : ℝ)| ≤ 1 := by cases int.units_eq_one_or r with hh hh; rw hh; simp
+
+/-- The CHSH inequality in `ℤˣ`. -/
+lemma CHSH_inequality_of_int_units (A₀ A₁ B₀ B₁ : ℤˣ) :
+  (A₀ : ℝ) * B₀ + A₀ * B₁ + A₁ * B₀ + (-A₁) * B₁ + -2 ≤ 0 :=
+  by cases pm_one_space_vals A₀ with hA0 hA0;
+    cases pm_one_space_vals A₁ with hA1 hA1;
+    cases pm_one_space_vals B₀ with hB0 hB0;
+    cases pm_one_space_vals B₁ with hB1 hB1;
+    rw [hA0, hA1, hB0, hB1]; ring_nf; simp
+
+end preliminaries
+
+section bells_inequality_1964
+
+variables {Ω : Type*} [measurable_space Ω] (ℙ : probability_measure Ω)
```

Comment:
Instead of declaring `ℙ : probability_measure Ω`, it is better to declare `ℙ : measure Ω` and a instance `is_probability_measure ℙ`. The type `probability_measure` is only used when we want to deal with convergence of measures.